AllHubs-June-2024-Liftoff/kavin-WHELness
# Controllers, endpoints, and axios service classes for CRUD functionality

# Controllers, endpoints, and axios service classes for CRUD functionality

## Description

This pull request includes:

- Controllers for each repository that include GET, POST, and DELETE endpoints for each repository object.  These controllers have logging (using LoggerFactory) for easier debugging.  They also include basic error handling.  Each controller method returns a `ResponseEntity` with an appropriate response type (`ok` for 200 responses with content, `noContent` for 204 responses that are successful but have no content, `notFound` for 404 failures).  

- A folder of axios service classes we can use in the frontend instead of directly hitting endpoints.  Right now, to get all recipes in a component, we would need to do something like this:
  ```javascript
  useEffect(() => {
          axios
              .get("http://localhost:8080/recipe/all")
              .then((response) => { 
                  setRecipes(response.data);
                  setLoading(false);
              })
              .catch((err) => {
                  setError(err.message);
                  setLoading(false);
              });
      }, []);
  ```

  This leaves room for error if we use the wrong base url, or don’t know the correct endpoint.  We would also have to do concatenation to use path variables (getting a recipe by ID, for instance).  
  With these service classes, you would simply import the one you want (in this case, `recipeService` from `src/services/recipeService.js`) and use `getAll()`.  

  ```javascript
  useEffect(() => {
          recipeService.getAll()
              .then((response) => { 
                  setRecipes(response.data);
                  setLoading(false);
              })
              .catch((err) => {
                  setError(err.message);
                  setLoading(false);
              });
      }, []);
  ```

  Any endpoint that takes a path variable now has a service method that just takes a parameter.  For instance: `recipeService.getById(1)` would create an axios call that looks like `axios.get("http://localhost:8080/recipe/" + id)`

* Added an alias to the frontend’s `src` folder so we don’t have to use relative paths.  If you want to import something from `src/components`, you can import it from `@/components`.  I got this idea from [here](https://dev.to/jumbo02/how-to-setup-path-alias-vite-react-5426), but seems like it is fairly common practice.
* I have also included extensive JSDoc documentation for each service class.  In IntelliJ and VSCode, you can hover over a service or method and it will tell you how to use it, what parameters it takes, and what it returns. I hope this helps make it easier to use our backend!  Examples:
![image](https://github.com/user-attachments/assets/25130d8b-d649-4a19-bc89-7ee77dee768d)
![image](https://github.com/user-attachments/assets/232944f5-05d7-451d-857e-2fa3755fab90)
* This pull request also _should_ fix the issue with IntelliJ and VSCode personal preference folders ending up in our repository. Once they are committed, adding them to our `.gitignore` won't remove them.  You have to do `git rm --cached -r .idea` and commit.  Hopefully this fixes it for everyone after this gets merged.

## Steps to Test

- Checkout my branch
- Start up the backend 
- In a browser, go to `http://localhost:8080/ingredient/all` - you should see a json response of all recipes in the database.  In your backend terminal window, should also see logging indicating what controller method is being used.
- Go to `http://localhost:8080/ingredient/1` - you should see a json response of the ingredient with id 1.  You should see some logging.
- Go to `http://localhost:8080/ingredient/123456` - you should get a 404, and you should see some error logs in the backend terminal window.
- Repeat these steps for any other endpoints you’d like to test.
